dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/DefaultAnalyticsService.java

```diff
@@ -255,10 +271,79 @@ private Grid getAggregatedDataValuesTableLayout(
 
     Map<String, Object> valueMap = AnalyticsUtils.getAggregatedDataValueMapping(grid);
 
+    List<List<DimensionalItemObject>> alternativeGridColumns =
+        reorderItems(getGridItems(grid, columnsDimensionItemsByDimension, columns));
+    List<List<DimensionalItemObject>> alternativeGridRows =
+        reorderItems(getGridItems(grid, rowsDimensionItemsByDimension, rows));
+
     return visualization.getGrid(
         new ListGrid(grid.getMetaData(), grid.getInternalMetaData()),
         valueMap,
+        alternativeGridColumns,
+        alternativeGridRows,
         params.getDisplayProperty(),
         false);
   }
+
+  private List<List<DimensionalItemObject>> reorderItems(
+      List<List<DimensionalItemObject>> alternateItems) {
+    return alternateItems.stream().sorted(this::compareItems).collect(Collectors.toList());
+  }
+
+  private int compareItems(List<DimensionalItemObject> dio, List<DimensionalItemObject> other) {
+    return dio.get(0).getDisplayName().compareTo(other.get(0).getDisplayName());
+  }
+
+  /**
+   * Returns alternative grid dimensional items based on the given grid and all dimension items.
+   * Alternative grid items are used improve the performance of the grid rendering when the
+   * combination of dimension items is large. The alternative grid items are a list of lists of
+   * dimension items where each list represents a possible combination of dimension items that are
+   * used to render the grid.
+   *
+   * @param grid the grid
+   * @param dimensionItemsByDimension a map of dimension items by dimension
+   * @param dimensionIds the dimension ids
+   * @return the alternative grid items
+   */
+  static List<List<DimensionalItemObject>> getGridItems(
+      Grid grid,
+      Map<String, List<DimensionalItemObject>> dimensionItemsByDimension,
+      List<String> dimensionIds) {
+
```

Comment:
Blank line

Reply:
I think sometimes a blank line makes the code more readable